ReadMe for out of process samples (#470)
Azure/azure-functions-sql-extension: Worker.Extension.Sql/README.md

# SQL Extension .NET Worker

> Welcome to the Sql Extension .NET Worker Repository. The .NET Worker provides .NET 6 support for SQL Bindings in Azure Functions, introducing an **Isolated Model**, running as an out-of-process language worker that is separate from the Azure Functions runtime. This allows you to have full control over your application's dependencies as well as other new features like a middleware pipeline. A .NET Isolated function app works differently than a .NET Core 3.1 function app. For .NET Isolated, you build an executable that imports the .NET Isolated language worker as a NuGet package. Your app includes a [`Program.cs`](vscode-file://vscode-app/c:/Users/makoripa/AppData/Local/Programs/Azure%20Data%20Studio/resources/app/out/vs/code/electron-browser/workbench/samples/samples-outofproc/Program.cs) that starts the worker.

Reviewer: Remove the local path hardcoded in this - and it shouldn't be a vscode-file URL either

## Samples

You can find samples on how to use Sql Extension .NET Worker under `samples\samples-outofproc` ([link](https://github.com/Azure/azure-functions-sql-extension/tree/main/samples/samples-outofproc)).

## Create and run .NET Isolated functions

**Note: Visual Studio and Visual Studio Code support is on the way. In the meantime, please use `azure-functions-core-tools` or the sample projects as a starting point.**  

### Install .NET 6.0

Download .NET 6.0 [from here](https://dotnet.microsoft.com/download/dotnet/6.0)

### Install the Azure Functions Core Tools

To download Core Tools, please check out our docs at [Azure Functions Core Tools](https://github.com/Azure/azure-functions-core-tools)

### Create a .NET Isolated project

In an empty directory, run `func init` and select `dotnet (Isolated Process)`.

### Add a function

Run `func new` and select `HttpTrigger` trigger. Fill in the function name.

### Configure function App

1. Get your SQL connection string
    
    Local SQL Server - Use this connection string, replacing the placeholder values for the database and password.  
      

>     `Server=localhost;Initial Catalog={db_name};Persist Security Info=False;User ID=sa;Password={your_password};` Azure SQL Server - Browse to the SQL Database resource in the \[Azure portal\](https://ms.portal.azure.com/)  

Reviewer: Fix the portal link - it looks like this currently since you're escaping the brackets

![image](https://user-images.githubusercontent.com/28519865/200739319-93a1cbe5-15c8-4673-8b17-091eec8805a5.png)

    \- In the left blade click on the **Connection Strings** tab  
    \- Copy the **SQL Authentication** connection string  
      
    (_Note: when pasting in the connection string, you will need to replace part of the connection string where it says '{your\_password}' with your Azure SQL Server password_)
2. Open the generated `local.settings.json` file and in the `Values` section verify you have the below. If not, add the below and replace `{connection_string}` with the your connection string from the previous step:
    
    ```
    "AzureWebJobsStorage": "UseDevelopmentStorage=true",
    "FUNCTIONS_WORKER_RUNTIME": "dotnet-isolated",
    "SqlConnectionString": "{connection_string}"
    
    ```
    

> 3. Verify your `host.json` looks like the below:

Reviewer: Is there any specific reason to include this step? There's nothing here that they need as far as I can see

    
    ```
    {
        "version": "2.0",
        "logging": {
            "applicationInsights": {
                "samplingSettings": {
                    "isEnabled": true,
                    "excludedTypes": "Request"
                }
            }
        }
    }
    
    ```
    
4. You have setup your local environment and are now ready to create your first out of proc SQL bindings! Continue to the [input](#Input-Binding-Tutorial) and [output](#Output-Binding-Tutorial) binding tutorials, or refer to [Samples](https://github.com/Azure/azure-functions-sql-extension/tree/main/samples/samples-outofproc) for information on how to use the bindings out of proc and explore on your own.
    

### Run functions locally

Run `func host start` in the sample function app directory.

## Tutorials

#### Input Binding Tutorial

> Note: This tutorial requires that a SQL database is setup as shown in [Create a SQL Server](#Create-a-SQL-Server).

Reviewer: This link doesn't lead anywhere

> - Open your app that you created in [Create a Function App](#create-a-function-app) in VSCode

Reviewer: ```suggestion
- Open your app that you created in [Create a Function App](#create-a-function-app) in VS Code
```

> - Open your app that you created in [Create a Function App](#create-a-function-app) in VSCode

Reviewer: This link doesn't lead anywhere

- Press 'F1' and search for 'Azure Functions: Create Function'
- Choose HttpTrigger -> (Provide a function name) -> Company.namespace -> anonymous
- In the file that opens, replace the `public static async Task<IActionResult> Run` block with the below code.

    ```csharp
    [Function("GetEmployees")]
    public static async Task<IActionResult> Run(
        [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post", Route = "employees")] HttpRequest req,
        ILogger log,

>         [SqlInput("select * from Employees",

Reviewer: ```suggestion
        [SqlInput("SELECT * FROM Employees",
```

        CommandType = System.Data.CommandType.Text,
        ConnectionStringSetting = "SqlConnectionString")]

>         IEnumerable<Employee> employee)

Reviewer: ```suggestion
        IEnumerable<Employee> employees)
```

    {

>         return new OkObjectResult(employee);

Reviewer: ```suggestion
        return new OkObjectResult(employees);
```

                    FirstName = "Hi",
                    LastName = "SQLupdate",
                    Company = "Microsoft",
                    Team = "Functions"
                },
            };

        return output;
    }
    ```

    *In the above, "dbo.Employees" is the name of the table our output binding is upserting into. The line below is similar to the input binding and specifies where our SqlConnectionString is. For more information on this, see the [Output Binding](https://github.com/Azure/azure-functions-sql-extension/blob/main/README.md#Output-Binding) section*

- Add 'using Microsoft.Azure.Functions.Worker.Extension.Sql;' for using *SqlOutput*, the out of proc sql output binding.
- Hit 'F5' to run your code. Click the link to upsert the output array values in your SQL table. Your upserted values should launch in the browser.
- Congratulations! You have successfully created your first out of proc SQL output binding! Checkout [Output Binding](https://github.com/Azure/azure-functions-sql-extension/blob/main/README.md#Output-Binding) for more information on how to use it and explore on your own!

## Differences from in-process bindings

> As stated in the functions [documentation](https://learn.microsoft.com/en-us/azure/azure-functions/dotnet-isolated-process-guide#output-bindings)

Reviewer: ```suggestion
As stated in the functions [documentation](https://learn.microsoft.com/azure/azure-functions/dotnet-isolated-process-guide#output-bindings)
```
